infra/token-transfer-server/src/tasks/transfer.js

```diff
@@ -0,0 +1,113 @@
+const fs = require('fs')
+const moment = require('moment')
+
+const Token = require('@origin/token/src/token')
+
+const {
+  discordWebhookUrl,
+  largeTransferThreshold,
+  largeTransferDelayMinutes,
+  watchdogPath,
+  networkId
+} = require('../config')
+const { postToWebhook } = require('../lib/webhook')
+const { Transfer, Sequelize } = require('../models')
+const { executeTransfer } = require('../lib/transfer')
+const logger = require('../logger')
+const enums = require('../enums')
+
+const watchdogFile = `${watchdogPath}/execute.pid`.replace(/\/\/+/g, '/')
+
+const initWatchdog = () => {
+  // Check there is no existing watchdog.
+  if (fs.existsSync(watchdogFile)) {
+    throw new Error(`Watchdog detected at ${watchdogFile}. Processing aborted.`)
+  }
+
+  // Create a watchdog for this run.
+  fs.writeFileSync(watchdogFile, `${process.pid}`)
+}
+
+const clearWatchdog = () => {
+  // Clean watchdog.
+  if (fs.existsSync(watchdogFile)) {
+    fs.unlinkSync(watchdogFile)
+  }
+}
+
+const executeTransfers = async () => {
+  logger.info('Running execute transfers job...')
+
+  initWatchdog()
+
+  const waitingTransfers = await Transfer.findAll({
+    where: {
+      [Sequelize.Op.or]: [
+        {
+          status: enums.TransferStatuses.WaitingConfirmation
+        },
+        {
+          status: enums.TransferStatuses.Processing
+        }
+      ]
+    }
+  })
+  if (waitingTransfers.length > 0) {
+    throw new Error(
+      `Found unconfirmed transfer(s). Fix before running this script again.`
+    )
+  }
+
+  const cutoffTime = moment.utc().subtract(largeTransferDelayMinutes, 'minutes')
+  const transfers = await Transfer.findAll({
+    where: {
+      [Sequelize.Op.or]: [
+        {
+          status: enums.TransferStatuses.Enqueued,
+          amount: { [Sequelize.Op.gte]: largeTransferThreshold },
+          createdAt: { [Sequelize.Op.lte]: cutoffTime }
+        },
+        {
+          status: enums.TransferStatuses.Enqueued,
+          amount: { [Sequelize.Op.lt]: largeTransferThreshold }
+        }
+      ]
+    }
+  })
+
+  logger.info(`Processing ${transfers.length} transfers`)
+
+  for (const transfer of transfers) {
+    logger.info(`Processing transfer ${transfer.id}`)
+    const result = await executeTransfer(transfer)
+    logger.info(
+      `Processed transfer ${transfer.id}. Status: ${result.txStatus} TxHash: ${result.txHash}`
+    )
+  }
+
+  clearWatchdog()
```

Comment:
What is the purpose of the watchdog specifically? Is it for:
A. Preventing more than 1 instance of the process to get started
B. Preventing the in-process cron scheduler to somehow start more than 1 job

If it's for A. maybe the watchdog should get initialized when the process starts and deleted at shutdown.

BTW, not sure if we have decided under which environment we will run the server.
 - If it's kubernetes, I'm not sure exactly about the detail of the lifecycle of a container. Does kube bring a new container and check its health before terminating the old one? If yes that could potentially cause 2 cron jobs to run in parallel.
 - Not sure about heroku
Overall I wonder if it would be simpler to use the DB as a central persistent storage for the watchdog. Like a new "watchdog" table.

Reply:
For B. More than 1 instance of the process is actually fine, but we really don't want more than one instance of the transfer processing job.

Re: kubernetes, it depends on the resource type. A StatefulSet will terminate one before starting the other, but a standard deployment will run into the problem you described.

Nothing decided yet, but a DB is probably more robust. You raise good points. I'm going to firm this up more later I think, but I want to test this out of on dev/staging before making any decisions.

Reply:
Sgtm!

Reply:
I moved the watchdog stuff into a separate SQL table which records task start and end times and links the transfers via a foreign key. Would you mind giving it a look? Then will merge this :tada:

Reply:
sure... will look a bit later tonite...